src/main/java/com/tvd12/gamebox/octree/OcTreeNode.java

```diff
@@ -0,0 +1,204 @@
+package com.tvd12.gamebox.octree;
+
+import com.tvd12.gamebox.entity.PositionAware;
+import com.tvd12.gamebox.math.Vec3;
+import lombok.Setter;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class OcTreeNode<T extends PositionAware> {
+    
+    @Setter
+    private OcTreeNode<T> parentNode = null;
+    private final int maxItems;
+    private final float minNodeSize;
+    private final OcBound bound;
+    private final Set<T> items = new HashSet<>();
+    private final List<OcTreeNode<T>> children = new ArrayList<>();
+    
+    private static final int NUM_CHILDREN = 8;
+
+    public OcTreeNode(OcBound bound, int maxItems, float minNodeSize) {
+        this.bound = bound;
+        this.maxItems = maxItems;
+        this.minNodeSize = minNodeSize;
+    }
+
+    public OcTreeNode<T> insert(T newItem, Vec3 position) {
+        if (!this.bound.containsPosition(position)) {
+            return null;
+        }
+
+        if (isLeaf()) {
+            if (this.items.size() < maxItems || this.bound.getMaxDimension() < 2 * minNodeSize) {
+                this.items.add(newItem);
+                return this;
+            }
+            createChildren();
+            passItemsToChildren();
+        }
+
+        return insertItemToChildren(newItem, position);
+    }
+
+    private void createChildren() {
+        for (int i = 0; i < NUM_CHILDREN; ++i) {
+            OcBound bound = this.bound.getOctant(OcLocation.of(i));
+            OcTreeNode<T> child = new OcTreeNode<>(bound, maxItems, minNodeSize);
+            this.children.add(child);
+            child.setParentNode(this);
+        }
+    }
+
+    private void passItemsToChildren() {
+        this.items.forEach(
+            item -> insertItemToChildren(item, item.getPosition())
+        );
+        this.items.clear();
+    }
+
+    private OcTreeNode<T> insertItemToChildren(T item, Vec3 position) {
+        for (int i = 0; i < NUM_CHILDREN; ++i) {
+            OcTreeNode<T> nodeContainingInsertedItem = this.children.get(i)
+                .insert(item, position);
+            if (nodeContainingInsertedItem != null) {
+                return nodeContainingInsertedItem;
+            }
+        }
+        return null;
+    }
+
+    public boolean remove(T item) {
+        if (!this.bound.containsPosition(item.getPosition())) {
+            return false;
+        }
+        if (isLeaf()) {
+            return removeItemFromThisLeaf(item);
+        }
+        return removeFromChildren(item);
+    }
+
+    private boolean removeItemFromThisLeaf(T item) {
+        if (!this.items.contains(item)) {
+            return false;
+        }
+        this.items.remove(item);
+        tryMergingChildrenOfParentNode();
+        return true;
+    }
+
+    private boolean removeFromChildren(T item) {
+        for (int i = 0; i < NUM_CHILDREN; ++i) {
+            boolean isPlayerRemoved = this.children.get(i)
```

Comment:
this.children.get(i) có thể null em nhỉ? anh nghĩ là vẫn check null ở đây cho an toàn, nếu ko test được bằng public thì dùng reflection để test private cũng được